src/python/TaskWorker/Actions/DagmanSubmitter.py

```diff
@@ -507,7 +507,7 @@ def submitDirect(self, schedd, cmd, arg, info): #pylint: disable=R0201
             for k, v in dagAd.items():
                 if k == 'X509UserProxy':
                     v = os.path.basename(v)
-                if isinstance(v, basestring):
+                if isinstance(v, str):
```

Comment:
Since my job is to be pedantic, if we wanted to have the same behavior as in py2 this should have been

```diff
- if isinstance(v, str):
+ if isinstance(v, (str, bytes)):
```

but is no problem was found, guaranteeing that this `v` is not going to be bytes, then I am happy! :)

Reply:
we do nor care for py2 compatibility here. Pure Python3. Is that enough to claim "strings will never be bytes" ? Or at least "if some new developer writes bytes where string is expected, let's raise an exception and abort immediately"

Reply:
> we do nor care for py2 compatibility here. Pure Python3

yeah, i understand this. the fact is that in py2 `basestring` meant both `str` and `unicode`, so if in py3 we want to keep the same behavior as py2, we would need to match both of those.

Moreover, i would align to the behavior of [classad.quote(input)](https://htcondor.readthedocs.io/en/latest/apis/python-bindings/api/classad.html#classad.quote)  (which is called if this `if` gives True) and `str.format()` (which is where the output of classad.quote() ends up):

```plaintext
>>> classad.quote("hello, 'world'!")
'"hello, \'world\'!"'
>>> classad.quote(b"hello, 'world'!")
'"hello, \'world\'!"'
>>> classad.quote("hello, 'ẘøґʟⅾ'!")
'"hello, \'\\341\\272\\230\\303\\270\\322\\221\\312\\237\\342\\205\\276\'!"'
>>> classad.quote(b"hello, 'ẘøґʟⅾ'!")
  File "<stdin>", line 1
SyntaxError: bytes can only contain ASCII literal characters.

>>> "hello {0}".format("world")
'hello world'
>>> "hello {0}".format(b"world")
"hello b'world'"

```

conclusions:
- it is better to make sure that we do not pass bytes to `.format()', if we do not want to risk having `b''` all around our strings -> if we use `if isinstance(v, str):`, then bytes will go straight to `.format()`
- `classad.quote()` happily accept both py3 `str` and `bytes` as input, returning py3 `str` ->
  a. if we are sure that `v` is unlikely to be bytes or it can be a bytes string that does not contain any non-ascii characters, i would go for `if isinstance(v, (str, bytes)):`
  b. if `v` can be a bytes strings that contains non-ascii characters, then i would use 

```python
if isinstance(v, (str, bytes)):
   value = classad.quote(decodeBytesToUnicode(v))
```

I do not have a final answer. as always, the solution depends on the type of our variables

Reply:
> Is that enough to claim "strings will never be bytes"

yeah, i mean, we can just claim this and hope `v` will never contain bytes. in case it does, we will have `b''` all around and we will (hopefully) know where they are from

> "if some new developer writes bytes where string is expected, let's raise an exception and abort immediately"

the current solution does not raise any exception, just lets the bytes go though formatting them with some `b''`. We can add a `if isinstance(v, bytes): raise ...`, but feels kind of aggressive. we can just convert those to unicode

Reply:
let's be concrete. That if only applies to the variables defined in the lines above in the same method (from lin 468 down) and used to do some magic to turn human readable classAds definitions into the precise quoting format required in a condor `subdag.ad` file. I.e. will only deal with things like `dagAd["CMS_SubmissionTool"] = "CRAB"` . Let's do whatever you like, but we should make sure we do not put `b'   '` in the classad file, I doubt condor will take it.

Reply:
Then i think `if isinstance(v, (str, bytes)):` will be a perfect fit for us here.

Reply:
OK. will be fixed in next commit